### PR TITLE
Fix a few small bugs in the plugin framework

### PR DESCRIPTION
**Pull request by emontnemery**
Fix a few small bugs in the plugin framework.
Add support to modify device settings.
Corrections of inbound and outbound MQTT PUBLISH packages.

### CODE REVIEW

**Comment by emontnemery** [comment]
Sorry about the number of commits, let me know if I should squash them.

File: hardware/plugins/PluginProtocols.cpp [hunk 1]
```diff
@@ -883,169 +889,202 @@ namespace Plugins {
...
 				}
-			}
-			else
+			} while ((encodedByte & 128) != 0);
+
+			if (iRemainingLength > std::distance(it, m_sRetainedData.end()))
 			{
-				ADD_INT_TO_DICT(pMqttDict, "Status", -1);
-				ADD_UTF8_TO_DICT(pMqttDict, "Description", std::string("Invalid message length"));
+				_log.Log(LOG_TRACE, "(%s) Too little data (got %u, expected %u).", __func__, std::distance(it, m_sRetainedData.end()), iRemainingLength);
```
Comment (gizmocuz) [inline hardware/plugins/PluginProtocols.cpp]: Is it possible to rephrase this to 'Not enough data received' instead of 'Too little data' ?

**Review by gizmocuz — COMMENTED** [review]
Thanks for this !!! Just one very minor request....

File: hardware/plugins/PluginProtocols.cpp (hunk 1 above)
Comment (emontnemery) [reply hardware/plugins/PluginProtocols.cpp]: Yes, sure!  Should I squash the commits as well?

**Comment by gizmocuz** [comment]
Thanks for the change! But are you sure the do->while(0) is needed ? Made a small test application and runs/compiles without issues:

```
#include <string>
#include <vector>

#define MQTT_STRING_W_LEN(vVector, sString)	\
 { \
  	int sLen = sString.length(); \
	for (int i = 0; i<sLen; i++) vVector.push_back(sString[i]); \
 }

int main()
{
	std::string szTest("Hello World!");
	std::vector<uint8_t> result;
	MQTT_STRING_W_LEN(result, szTest);
    return 0;
}

```

I would favor that your macro's follow the standard parameter method of Input/Output as now it is Output/Input

**Comment by gizmocuz** [comment]
length() returns a size_t object...

**Comment by gizmocuz** [comment]
I am not 100% in the code, but to convert a string to a vector you could simple use

`std::vector<uint8_t> result2(szTest.begin(), szTest.end());`

File: hardware/plugins/PluginProtocols.cpp (hunk 1 above)
Comment (gizmocuz) [reply hardware/plugins/PluginProtocols.cpp]: No need, i could smash it as we merge ;)

**Comment by emontnemery** [comment]
The purpose of the do { .. } while (0) is to ensure the macro can be used the same way a function call with no special cases. Here's a good summary explaining why it's necessary: https://gcc.gnu.org/onlinedocs/cpp/Swallowing-the-Semicolon.html

Agreed on the other comments, I'll update.

Edit:
Changed to functions instead of macros because macro expansion means multiple instances of passed string literals will be constructed, so the begin() and end() iterators will actually refer to different strings.

**Comment by gizmocuz** [comment]
Thanks!
